---
title: "PredictioR: A Package for Biomarker Discovery in Immunotherapy Response"
author:
- name: Farnoosh Abbas-Aghababazadeh
  affiliation:
  - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
- name: Yacine Bareche
  affiliation:
  - &pm Faculty of Pharmacy, Université de Montréal, Montreal, Canada
  - &mbp Centre de Recherche du Centre Hospitalier de l’Université de Montréal, Institut du Cancer de Montréal, Montreal, Canada
- name: Minoru Nakano
  affiliation:
    - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
- name: John Stagg
  affiliation:
  - &pm Faculty of Pharmacy, Université de Montréal, Montreal, Canada
  - &mbp Centre de Recherche du Centre Hospitalier de l’Université de Montréal, Institut du Cancer de Montréal, Montreal, Canada
- name: Benjamin Haibe-Kains
  affiliation:
    - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
    - &mbp Department of Medical Biophysics, University of Toronto, Toronto, Canada
  email: benjamin.haibe.kains@utoronto.ca
#package: PredictioR
output:
  BiocStyle::html_document
abstract: |
  Description of your vignette
vignette: |
  %\VignetteIndexEntry{PredictioR: A Package for Biomarker Discovery in Immunotherapy Response}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The PredictioR package provides relevant functions for biomarker discovery in immunotherapy response through pan-cancer and cancer-specific analyses. This package includes a number of algorithms for predicting immunotherapy response. The package also includes implementations of immunotherapy signature score computation.

Please refer to the manuscript URL (https://pubmed.ncbi.nlm.nih.gov/36055464/).

# Loading Functions

First we load the functions into the workspace. To compute gene associations with immunotherapy response, calculate signature scores, assess signature associations with immunotherapy response, integrate the estimates, and compare them, we need to load all the necessary packages

```{r load, echo=TRUE, message=FALSE, warning=FALSE}

#devtools::install_github("bhklab/PredictioR")
library(PredictioR)
library(survcomp)
library(GSVA)
library(dplyr)
library(meta)
library(forestplot)
library(ggplot2)
library(ggrepel)
library(gridExtra)
library(data.table)
library(kableExtra)
library(summarytools)
library(MultiAssayExperiment)

```


# Load Immunotherapy Datasets 

The following clinical multimodal immunotherapy datasets are publicly available on the [GitHub](https://github.com/bhklab/ClinSets.git) data repository. Datasets are also available from the cloud-based platform [ORCESTRA](https://www.orcestra.ca/clinical_icb), ensuring that the curation and processing of the clinical genomic data are fully transparent and reproducible. These datasets are used in biomarker discovery for immunotherapy response through pan-cancer, cancer-specific, and treatment-specific analyses.

|Dataset      |Patients [\#]  |Cancer type   |RECIST criteria |Clinical endpoints |Molecular data |PMID  |      
|-------------|-------------  |------------- |--------------|--------------|----------|----------|------|
| Hugo        | 27            | Melanoma     | Available    | OS       |RNA       | [26997480](https://pubmed.ncbi.nlm.nih.gov/26997480/)| 
| Liu         | 121           | Melanoma     | Available    | PFS/OS   |RNA/DNA   | [31792460](https://pubmed.ncbi.nlm.nih.gov/31792460/)| 
| Miao1       | 33            | Kidney       | Available    | PFS/OS   |RNA/DNA   | [29301960](https://pubmed.ncbi.nlm.nih.gov/29301960/)|
| Nathanson   | 24            | Melanoma     | Available    | OS       |RNA/DNA   | [27956380](https://pubmed.ncbi.nlm.nih.gov/27956380/)|
| Padron      | 45            | Pancreas     | Available    | PFS/OS   |RNA       | [35662283](https://pubmed.ncbi.nlm.nih.gov/35662283/)|
| Riaz        | 46            | Melanoma     | Available    | OS       |RNA/DNA   | [29033130](https://pubmed.ncbi.nlm.nih.gov/29033130/)|
| Van_Allen   | 42            | Melanoma     | Available    | PFS/OS   |RNA/DNA   | [26359337](https://pubmed.ncbi.nlm.nih.gov/26359337/)|
| Mariathasan | 195           | Bladder      | Available    | OS       |RNA/DNA   | [29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/)| 
: (\#tab:table1) Detailed overview for the immunotherapy datasets.

The immunotherapy dataset is stored in a SummarizedExperiment object. For RNA profiles, we utilize log2-transformed TPM (Transcripts Per Million) data from protein-coding genes. For each dataset, genes with zero expression values in at least 50\% of the samples are filtered out. Note that only studies with at least 20 patients are included in the following analyses.

As an example, the Mariathasan immunotherapy dataset with PMID [29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/) includes RNA expression, clinical characteristics, and gene metadata. A total of 195 patients have both clinical data and RNA expression data. Additionally, the Mariathasan dataset contains 17,993 protein-coding genes. Due to size limitations, only a subset of the ICB data was shared.

```{r load Mariathasan data}

# gene expression, clinical characteristics, and gene meta data 
expr <- assay(ICB_small_Mariathasan)
clin <- colData(ICB_small_Mariathasan) %>% as.data.frame()
annot <- rowData(ICB_small_Mariathasan) %>% as.data.frame()

# number of protein-coding genes and patients
dim(expr)

```

The detailed clinical characteristics of the Mariathasan dataset including cancer type, age, sex, response (R vs. NR), and overall survival (OS). The summary table represents that in the Mariathasan dataset about 32\% of the response data is missing, while no missing for survival variables. Additionally, almost 79\% of patients are male. Note that "R" and "NR" represent responder and non-responder, respectively.  

```{r Braun clinical data summary}

# summary table of selected clinical variable 
sub_clin <- clin[, c("age","sex", "response","survival_time_os")]
print(dfSummary(sub_clin, 
                style = "grid", 
                plain.ascii = FALSE, 
                graph.col = FALSE), 
      method = 'render')

```


# Load Immunotherapy RNA Signatures 

We evaluate the reproducibility of a compendium of more than 50 RNA signatures described as immunotherapy biomarkers. The following published immunotherapy RNA signatures are available on the [GitHub](https://github.com/bhklab/SignatureSets) data repository. To compute the signatures' scores, different methods are used, including weighted mean, single sample gene set enrichment analysis (ssGSEA), Gene Set Variation Analysis (GSVA) [PMID 23323831](https://pubmed.ncbi.nlm.nih.gov/23323831/), and specific computational algorithms.

# Biomarkers and Immunotherapy Response Association

Association of specific biomarkers with immunotherapy response (R vs NR) is assessed using a logistic regression model. Association of specific biomarkers with immunotherapy survival (either PFS or OS) is assessed using Cox proportional hazards model to assess the discrimination or separation of a survival model. 

To assess the association between a biomarker and immunotherapy response (R vs NR), the logistic regression model is given as

$$\textit{logit}(E(Y|X)) = \textit{logit}(p) = \textit{ln} (\frac{p}{1-p}) = \beta_0 + \beta_1 X + \epsilon$$
where

* $Y$ denotes the outcome or dependent variable (e.g., immunotherapy response); this is a binary variable
* $X$ denotes the predictor of interest or the independent variable (e.g., gene expression)
* $p$ denotes the probability of the event occurring, and $\frac{p}{1-p}$ shows the odds ratio (OR)
* $\beta_0$ denotes the Y-intercept when X is zero; this is not informative for logistic regression models
* $\beta_1$ denotes the slope or the expected change in log odds per unit change in X
* $\epsilon$ denotes the random errors follow normal distribution

To assess the association between a biomarker and survival immunotherapy (e.g., OS or PFS), we fit the Cox proportional hazards (PH) model. The semi-parametric Cox PH model is given as 

$$h(t) = h_0(t) \times \text{exp}(\beta X)$$
where 

* $t$ represents the survival time.
* $h(t)$ is the hazard function determined by a set of covariate (e.g., X or gene expression)
* The coefficient $\beta$ measures the impact of covariate
* The term $h_0$ is called the baseline hazard. It corresponds to the value of the hazard if the covariate is equal to zero. The $t$ in $h(t)$ reminds us that the hazard may vary over time
* The quantity $\text{e}^\beta$ is called hazard ratios (HR)

When needed p-values are corrected for multiple testing using the Benjamini-Hochberg (False Discovery Rate, FDR) method. Associations are deemed statistically significant for p-values and/or FDR lower or equal to 5\%.

Note that Coef represents the log hazard ratio (logHR) or log odds ratio (logOR) depending on the analysis. 

## Association of Gene (CXCL9) with Immunotherapy Overall Survival Response 

We limit the association analysis of CXCL9 to a selected bladder cancer dataset (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/)). We assess the association of the CXCL9 gene with overall survival (OS) in response to immunotherapy, to gain more insight into the mechanisms of response and resistance to immunotherapy. The results show that CXCL9 is significantly associated with low survival (logHR = -3.47, p = 3e-05). Note that HR represents the hazard ratio in log scale.

```{r gene association os CXCL9, message=FALSE}

## Association of CXCL9 gene with OS for selected dataset Mariathasan Bladder

dat_expr <- as.matrix(assay(ICB_small_Mariathasan))
dat_clin <- colData(ICB_small_Mariathasan) %>% as.data.frame()
cancer_type <- names( table( dat_clin$cancer_type )[ table( dat_clin$cancer_type ) >= 15 ] )
       
survCont( status = dat_clin$event_occurred_os[ dat_clin$cancer_type %in% cancer_type ] ,
          time = dat_clin$survival_time_os[ dat_clin$cancer_type %in% cancer_type ] ,
          time.censor = 36 , 
          var = as.numeric( scale( dat_expr["CXCL9" , ] )))
     
```

## Association of Dichotomous Gene (CXCL9) with Immunotherapy Overall Survival Response  

We limit the association analysis of dichotomous CXCL9 to a selected bladder cancer dataset (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/)). The default median cutoff is applied to stratify the continuous expression to the High and Low groups. We assess the association of CXCL9 gene with immunotherapy OS, to gain more insight into the mechanism of response and resistance to immunotherapy response. The results show that CXCL9 is significantly associated with low survival (logHR = -0.61, p = 7.16e-04). Note the HR represents the logHR.    


```{r gene association Os CXCL9 dicho, message=FALSE}

dat_expr <- as.matrix(assay(ICB_small_Mariathasan))
dat_clin <- colData(ICB_small_Mariathasan) %>% as.data.frame()
cancer_type <- names( table( dat_clin$cancer_type )[ table( dat_clin$cancer_type ) >= 15 ] )
       
survDicho( status = dat_clin$event_occurred_os ,
           time = dat_clin$survival_time_os,
           time.censor= 36,
           var = as.numeric( scale( dat_expr["CXCL9" , ] )),
           n0.cutoff = 5,
           n1.cutoff = 5,
           method = "median",
           var.type = FALSE)

```

### Kaplan-Meier Plot

The log-rank test is the most widely non-parametric used method of comparing two survival curves can be applied to assess difference in survival between the two Low and High groups.

```{r Fig1, message=FALSE, fig.width=5, fig.height=5, fig.cap = "CXCL9 association with immunotherapy OS", fig.align="center"}
  
KMPlot(status = dat_clin$event_occurred_os,
       time = dat_clin$survival_time_os,
       time.censor = 36,
       var =  as.numeric(scale(dat_expr["CXCL9", ])), 
       title = "CXCL9 and OS Association",
       xlab = "Time (Months)",
       ylab = "Overall Survival", 
       n0.cutoff = 5, 
       n1.cutoff = 5,
       var.type = FALSE)

```


## Association of Genes with Immunotherapy Overall Survival Response

We assess the association of genes in PredictIO Bareche RNA signature [36055464](https://pubmed.ncbi.nlm.nih.gov/36055464/) with immunotherapy OS, to gain more insight into the mechanism of response and resistance to immunotherapy response. 

```{r gene association os , message=FALSE}

genes <- PredictIO_Bareche$gene_name
cox_os <- geneSurvCont(dat.icb = ICB_small_Mariathasan,
                       time.censor = 36,
                       missing.perc = 0.5,
                       const.int = 0.001,
                       n.cutoff = 15,
                       feature = genes,
                       study = 'ICB_Mariathasan',
                       surv.outcome = 'OS',
                       cancer.type = 'Bladder',
                       treatment = 'PD-1/PD-L1')
     
cox_os$FDR <- p.adjust(cox_os$Pval, method = "BH")
knitr::kable(head(cox_os[order(cox_os$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()

```

The association results for a selected Melanoma (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/)) dataset show that 40\% of genes are associated with immunotherapy survival with $FDR < 0.05$, where about 95\% of them are associated with worse survival.     
     
```{r Fig2, message=FALSE, fig.width=5, fig.height=5, fig.cap = "gene association with immunotherapy overall survival", fig.align="center"}
     
p <- volcanoPlot(feature = cox_os$Gene, 
                 coef = cox_os$Coef, 
                 pval = cox_os$Pval, 
                 padj = cox_os$FDR, 
                 neg.cutoff = 38, 
                 pos.cutoff = 2, 
                 x.lab = "logHR estimate",
                 padj.label = FALSE,
                 cutoff = 0.05,
                 colors = c("#EF8A62","#67A9CF", "#999999"),
                 coef.cutoff.up = 0,
                 coef.cutoff.down = 0)
     
p
     
```
     
Additionally, we assess the association of genes in PredictIO signature with immunotherapy response (R vs NR).
     
```{r gene association response, message=FALSE}
     
## identify genes' association with binary response (R vs NR)
logreg <- geneLogReg(dat.icb = ICB_small_Mariathasan,
                     missing.perc = 0.5,
                     const.int = 0.001,
                     n.cutoff = 15,
                     feature = genes,
                     study = 'ICB_Mariathasan', 
                     n0.cutoff = 10,
                     n1.cutoff = 10,
                     cancer.type = 'Bladder',
                     treatment = 'PD-1/PD-L1')
     
logreg$FDR <- p.adjust(logreg$Pval, method = "BH")
knitr::kable(head(logreg[order(logreg$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:50%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```
     
The association results for a selected Bladder (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/)) show that 35\% of genes are associated with immunotherapy response with $p <0.05$, while 26\% of them are associated with better respond.    
     
```{r Fig3, message=FALSE, fig.width=5, fig.height=4, fig.cap = "gene association with immunotherapy response (responder vs non-responder)", fig.align="center"}
     
p <- volcanoPlot(feature = logreg$Gene, 
                 coef = logreg$Coef, 
                 pval = logreg$Pval, 
                 padj = logreg$FDR, 
                 neg.cutoff = 26, 
                 pos.cutoff = 9, 
                 x.lab = "logOR estimate",
                 padj.label = FALSE,
                 cutoff = 0.05,
                 colors = c("#EF8A62","#67A9CF", "#999999"),
                 coef.cutoff.up = 0, 
                 coef.cutoff.down = 0)
     
p
     
```

### Aggregating Associations through Meta-analysis (Pan-cancer)
     
To improve reproducibility, the results of individual independent datasets are pooled using random-effects meta-analysis with inverse variance weighting in \textit{DerSimonian and Laird} random-effects models. Heterogeneity across studies is evaluated using the $Q$ statistic along with the $I^2$ index, which describes the proportion of total variation across studies attributable to heterogeneity rather than sampling error. Note that an $I^2$ value of $>50\%$ along with Cochran’s  $Q$ statistic $p < 0.05$ indicates moderate-to-high heterogeneity.

As an example, for the gene CXCL9, to generalize the association with immunotherapy survival, we apply a meta-analysis approach to integrate findings across datasets for pan-cancer analysis. The results show that the gene is significantly associated with worse survival ($\text{logHR} = -0.26, p = 2.9e-05$) with no heterogeneity across datasets ($I^2 = 0\%$, $Q$ $p = 0.45$).  
     
```{r meta-analysis pan-cancer}

expr <- list('ICB_Liu' = ICB_small_Liu, 'ICB_Padron' = ICB_small_Padron, 'ICB_Hugo' = ICB_small_Hugo, 
              'ICB_Mariathasan' = ICB_small_Mariathasan, 'ICB_Nathanson' = ICB_small_Nathanson, 
              'ICB_Riaz' = ICB_small_Riaz, 'ICB_Miao' = ICB_small_Miao, 'ICB_Van_Allen' = ICB_small_Van_Allen)
 
cancer_type <- c('Melanoma', 'Pancreas', 'Melanoma', 'Bladder', 'Melanoma', 'Melanoma', 'Kidney', 'Melanoma')
treatment_type <- c('PD-1/PD-L1', 'PD-1/PD-L1', 'PD-1/PD-L1', 'PD-1/PD-L1', 'CTLA4', 'IO+combo', 'PD-1/PD-L1', 'CTLA4')
 
assoc.res <- lapply(1:length(expr), function(k){

  geneSurvCont(dat.icb = expr[[k]],
               time.censor = 36,
               missing.perc = 0.5,
               const.int = 0.001,
               n.cutoff = 15,
               feature = 'CXCL9',
               study = names(expr)[k],
               surv.outcome = 'OS',
               cancer.type = cancer_type[k],
               treatment = treatment_type[k])
  
 })
assoc.res <- do.call(rbind, assoc.res)
    
# meta-analysis for a gene across datasets
res_meta <- metafun(coef = assoc.res$Coef, 
                    se = assoc.res$SE,
                    study  = assoc.res$Study, 
                    pval = assoc.res$Pval, 
                    n = assoc.res$N,
                    cancer.type = assoc.res$Cancer_type,
                    treatment = assoc.res$Treatment,
                    feature = "CXCL9", 
                    cancer.spec = FALSE, 
                    treatment.spec = FALSE)
     
# meta-analysis results
res_meta$meta_output
     
# summary of meta-analysis results
knitr::kable(res_meta$meta_summery, format="html", 
                  table.attr = "style='width:50%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```
     
The following forest plot displays the aggregated results across datasets for pan-cancer analysis. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed.  Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.
     
```{r Fig4, fig.width=8, fig.height=8, fig.cap ="Forest plot CXCL9 and OS association: Pan-cancer analysis", fig.align= 'center'}
     
forestPlot(coef = assoc.res$Coef, 
           se = assoc.res$SE, 
           study  = assoc.res$Study, 
           pval = assoc.res$Pval, 
           n = assoc.res$N, 
           cancer.type = assoc.res$Cancer_type,
           treatment = assoc.res$Treatment,
           feature = "CXCL9", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```
     
### Aggregating Associations through Meta-analysis (Per-cancer)

For cancer-specific analysis, consider meta-analysis when there are at least 3 datasets. The following forest plot displays the aggregated results across datasets as cancer-specific analysis. CXCL9 is associated with worse overall survival ($\text{logHR} = -0.24, p = 0.02$) using Melanoma datasets. 

```{r meta-analysis per-cancer}

# cancer specific meta-analysis for a gene across datasets
res_meta <- metaPerCanfun(coef = assoc.res$Coef, 
                          se = assoc.res$SE,
                          study  = assoc.res$Study, 
                          pval = assoc.res$Pval, 
                          n = assoc.res$N,
                          cancer.type = assoc.res$Cancer_type,
                          treatment = assoc.res$Treatment,
                          feature = "CXCL9", 
                          cancer.spec = TRUE)
     
# summary of meta-analysis results for each cancer types
knitr::kable(rbind(res_meta$Melanoma$meta_summery,
                   res_meta$Other$meta_summery), format="html", 
                   table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across Melanoma datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.


Subgroup meta-analysis analysis is considered to assess the impact of cancer type on the source of moderate-to-high heterogeneity. 

```{r Fig5, fig.width=8, fig.height=10, fig.cap ="Forest plot CXCL9 and OS association: Cancer-specific analysis", fig.align= 'center'}

forestPlotPerCan(coef = assoc.res$Coef, 
                 se = assoc.res$SE, 
                 study  = assoc.res$Study, 
                 pval = assoc.res$Pval, 
                 n = assoc.res$N, 
                 cancer.type = assoc.res$Cancer_type,
                 treatment = assoc.res$Treatment,
                 feature = "CXCL9", 
                 xlab = "logHR estimate", 
                 label = "logHR")
     
```

### Aggregating Associations through Meta-analysis (Per-treatment)

For treatment-specific analysis, consider meta-analysis when there are at least 3 datasets. The following forest plot displays the aggregated results across datasets as cancer-specific analysis. CXCL9 is associated with worse overall survival ($\text{logHR} = -0.19, p = 0.04$) using datasets targeted PD-1 and PD-L1. 

```{r meta-analysis per-treatment}

# treatment specific meta-analysis for a gene across datasets
res_meta <- metaPerTreatmentfun(coef = assoc.res$Coef, 
                                se = assoc.res$SE,
                                study  = assoc.res$Study, 
                                pval = assoc.res$Pval, 
                                n = assoc.res$N,
                                cancer.type = assoc.res$Cancer_type,
                                treatment = assoc.res$Treatment,
                                feature = "CXCL9", 
                                treatment.spec = TRUE)
     
# summary of meta-analysis results for each treatment types
knitr::kable(rbind(res_meta$`PD-1/PD-L1`$meta_summery,
                   res_meta$Other$meta_summery), format="html", 
                   table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across Melanoma datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.

Subgroup meta-analysis analysis is considered to assess the impact of cancer type on the source of moderate-to-high heterogeneity. 

```{r Fig6, fig.width=8, fig.height=10, fig.cap ="Forest plot CXCL9 and OS association: Treatment-specific analysis", fig.align= 'center'}

forestPlotPerTreatment(coef = assoc.res$Coef, 
                       se = assoc.res$SE, 
                       study  = assoc.res$Study, 
                       pval = assoc.res$Pval, 
                       n = assoc.res$N, 
                       cancer.type = assoc.res$Cancer_type,
                       treatment = assoc.res$Treatment,
                       feature = "CXCL9", 
                       xlab = "logHR estimate", 
                       label = "logHR")
     
```

## Signature Score Computation
     
RNA expression signatures, including genes without any assigned weights, are computed using the Gene Set Variation Analysis (GSVA) enrichment score with the \textit{GSVA} R package [REF]. RNA expression signatures containing genes with specific weights (+1 for up-regulated and -1 for down-regulated genes) are computed using the weighted mean expression approach. For RNA immunotherapy signatures that require specific computational algorithms, such as PredictIO, the original publication methods are used to obtain the signature scores.
     
For each dataset, RNA signatures are computed only if at least 80\% of their genes are present in the dataset. We also apply \textit{z}-score transformation to the genes of each RNA signature before the computation. The \textit{z}-score transformation is applied before GSVA or the weighted mean computation, and after the computation of the specific immunotherapy signatures.
     
### GSVA signature score computation
     
\textit{CYT_Rooney} ([PMID 25594174](https://pubmed.ncbi.nlm.nih.gov/25594174/)) is a RNA signature capturing the local immune cytolytic (CYT) activity. To get the signature score, the GSVA approach is applied. 
     
```{r signature score computation GSVA, message=FALSE, warning = FALSE}
     
geneSigScore <- geneSigGSVA(dat.icb = ICB_small_Mariathasan, 
                            sig = CYT_Rooney,
                            sig.name = "CYT_Rooney",
                            missing.perc = 0.5,
                            const.int = 0.001,
                            n.cutoff = 15,
                            sig.perc = 0.8, 
                            study = 'ICB_Mariathasan')
     
head(geneSigScore)
     
```
     
### Weighted mean signature score computation
     
\textit{EMT_Thompson} ([PMID 31683225](https://pubmed.ncbi.nlm.nih.gov/31683225/)) is a RNA signature capturing the epithelial-to-mesenchymal transition (EMT) activity. This EMT signature was a predictor of the failure of PD-(L)1 blockade in lung cancer patients. To get the signature score, the weighted mean approach is applied. 
     
```{r signature score computation weighted mean, message=FALSE, warning = FALSE}
     
geneSigScore <- geneSigMean(dat.icb = ICB_small_Mariathasan, 
                            sig = EMT_Thompson,
                            sig.name = 'EMT_Thompson',
                            missing.perc = 0.5,
                            const.int = 0.001,
                            n.cutoff = 15,
                            sig.perc = 0.8, 
                            study = 'ICB_Mariathasan')
     
head(geneSigScore)
    
```
     
### Specific algorithm (PredictIO) signature score computation
     
\textit{PredictIO} ([PMID 36055464](https://pubmed.ncbi.nlm.nih.gov/36055464/)) is a RNA signature that demonstrated better and more consistent ability to predict immunotherapy response as compared to other signatures. The specific algorithm is applied to computed the signature score. The PredictIO signature was identified by integrating gene expression data from a large pan-cancer discovery immunotherapy datasets. The validation immunotherapy datasets were used to assess the performance of the PredictIO compared with other published RNA immunotherapy signatures.
     
     
```{r signature score computation specific, message=FALSE, warning = FALSE}
  
geneSigScore <- geneSigPredictIO(dat.icb = ICB_small_Padron, 
                                 sig = PredictIO_Bareche,
                                 sig.name = 'PredictIO_Bareche',
                                 missing.perc = 0.5,
                                 const.int = 0.001,
                                 n.cutoff = 15,
                                 sig.perc = 0.8, 
                                 study = 'ICB_Padron')
     
head(geneSigScore)
     
```
     
     
## Association of RNA Signatures with Immunotherapy Response
     
For \textit{CYT_Rooney} signature, we assess the association of signature with immunotherapy OS. 
    
```{r signature score association with immunotherapy os, message=FALSE, warning = FALSE}

geneSig_os <- lapply(1:length(expr), function(k){
       
       geneSig <- geneSigGSVA(dat.icb = expr[[k]], 
                              sig = CYT_Rooney,
                              sig.name = 'CYT_Rooney',
                              missing.perc = 0.5,
                              const.int = 0.001, 
                              n.cutoff = 15,
                              sig.perc = 0.8, 
                              study = names(expr)[k])
       
       if(sum(!is.na(geneSig)) > 0){
         
         res <- geneSigSurvDicho(dat.icb = expr[[k]],
                                 geneSig = geneSig,
                                 time.censor = 24,
                                 n.cutoff = 15,
                                 study =  names(expr)[k],
                                 surv.outcome = "OS",
                                 n0.cutoff = 5,
                                 n1.cutoff = 5,
                                 sig.name = 'CYT_Rooney',
                                 method = 'median',
                                 var.type = FALSE,
                                 cancer.type = cancer_type[k],
                                 treatment = treatment_type[k])
         
       }else{
         
         res <- data.frame( Outcome = "OS",
                            Gene = NA, 
                            Study = names(expr)[k],
                            Coef = NA,
                            SE = NA,
                            N = NA,
                            Pval = NA,
                            Cancer_type= NA,
                            Treatment = NA) 
       }
       
       rownames(res) <- NULL
       
       res
       
     })
     
geneSig_os <- do.call(rbind, geneSig_os)
geneSig_os$FDR <- p.adjust(geneSig_os$Pval, method = "BH")
       
# top association results
knitr::kable(geneSig_os[order(geneSig_os$Pval, decreasing = FALSE), ], format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
     
```

### Aggregating Associations through Meta-analysis (Pan-cancer)

The associations across datasets are aggregated using the meta-analysis approach. The signature is negatively associated with OS ($logHR = -0.39, p = 0.05$) with heterogeneity across datasets ($I^2 = 45\%$, $Q$ $p = 0.08$).    
    
```{r signature score meta-analysis pan, message=FALSE, warning = FALSE }
     
res <- metafun(coef = geneSig_os$Coef, 
               se = geneSig_os$SE, 
               study  = geneSig_os$Study, 
               pval = geneSig_os$Pval, 
               n = geneSig_os$N, 
               cancer.type = geneSig_os$Cancer_type,
               treatment = geneSig_os$Treatment,
               feature = unique(geneSig_os$Gene),
               cancer.spec = FALSE,
               treatment.spec = FALSE)
     
     
# summary of meta-analysis
knitr::kable(res$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```
     
The forest plot shows the pan-cancer analysis. 
     
```{r Fig7, message=FALSE, warning = FALSE, fig.width=8, fig.height=5, fig.cap ="Forest plot (CYT_ROOney signature): Pan-cancer analysis", fig.align= 'center'}
     
forestPlot(coef = geneSig_os$Coef, 
           se = geneSig_os$SE, 
           study  = geneSig_os$Study, 
           pval = geneSig_os$Pval, 
           n = geneSig_os$N, 
           cancer.type = geneSig_os$Cancer_type,
           treatment = geneSig_os$Treatment,
           feature = unique(geneSig_os$Gene), 
           xlab = "logHR estimate", 
           label = "logHR")
     
     
```
     

# References


# SessionInfo

```{r sessionInfo, echo=FALSE}
#sessionInfo()
```

