---
title: "PredictioR: A Package for Biomarker Discovery in Immunotherapy Response"
author:
- name: Farnoosh Abbas-Aghababazadeh
  affiliation:
  - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
- name: Yacine Bareche
  affiliation:
  - &pm Faculty of Pharmacy, Université de Montréal, Montreal, Canada
  - &mbp Centre de Recherche du Centre Hospitalier de l’Université de Montréal, Institut du Cancer de Montréal, Montreal, Canada
- name: Minoru Nakano
  affiliation:
    - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
- name: John Stagg
  affiliation:
  - &pm Faculty of Pharmacy, Université de Montréal, Montreal, Canada
  - &mbp Centre de Recherche du Centre Hospitalier de l’Université de Montréal, Institut du Cancer de Montréal, Montreal, Canada
- name: Benjamin Haibe-Kains
  affiliation:
    - &pm Bioinformatics and Computational Genomics Laboratory, Princess Margaret Cancer Center,University Health Network, Toronto, Ontario, Canada
    - &mbp Department of Medical Biophysics, University of Toronto, Toronto, Canada
  email: benjamin.haibe.kains@utoronto.ca
#package: PredictioR
output:
  BiocStyle::html_document
abstract: |
  Description of your vignette
vignette: |
  %\VignetteIndexEntry{PredictioR: A Package for Biomarker Discovery in Immunotherapy Response}
  %\VignetteEncoding{UTF-8}
---

# Introduction

The PredictioR package provides relevant functions for biomarker discovery in immunotherapy response through pan-cancer and cancer-specific analyses. This package includes a number of algorithms for predicting immunotherapy response. The package also includes implementations of immunotherapy signature score computation.

Please refer to the manuscript URL (https://pubmed.ncbi.nlm.nih.gov/36055464/).

# Loading Functions

First we load the functions into the workspace. To compute gene associations with immunotherapy response, calculate signature scores, assess signature associations with immunotherapy response, integrate the estimates, and compare them, we need to load all the necessary packages

```{r load, echo=TRUE, message=FALSE}

library(stringr)
app_dir <- str_split(rstudioapi::getActiveDocumentContext()$path,'PredictioR')[[1]][1]

library(survcomp)
library(GSVA)
library(dplyr)
library(meta)
library(metafor)
library(forestplot)
library(ggplot2)
library(ggrepel)
library(gridExtra)
library(data.table)
library(kableExtra)
library(summarytools)
library(MultiAssayExperiment)

```


# Load Immunotherapy Datasets 

The following clinical multimodal immunotherapy datasets are publicly available on [GitHub](https://github.com/bhklab/ClinSets.git) data repository. Datasets are also obtained from the a cloud-based [ORCESTRA](https://www.orcestra.ca/clinical_icb) platform to ensure that the curation and processing of the clinical genomic data are fully transparent and reproducible. The following datasets are used in the biomarker discovery for immunotherapy response via pan-cancer and cancer-specific analyses. 

```{r load_data, eval=TRUE, results='hide', message=FALSE}

app_dir <- str_split(rstudioapi::getActiveDocumentContext()$path,'PredictioR')[[1]][1]
dir <- file.path(app_dir, 'PredictioR', "data/ICB") 

```

|Dataset      |Discovery / Validation |Patients [\#]  |Cancer type   |RECIST criteria |Clinical endpoints |Molecular data |PMID  |      
|-------------|-------------|---------------|--------------|--------------|----------|----------|------|
| Braun       | Discovery  | 181           | Kidney       | Available    | PFS/OS   |RNA/DNA   | [32472114](https://pubmed.ncbi.nlm.nih.gov/32472114/)|
| Gide        | Validation | 41            | Melanoma     | Available    | PFS/OS   |RNA       | [30753825](https://pubmed.ncbi.nlm.nih.gov/30753825/)| 
| Hugo        | Discovery  | 27            | Melanoma     | Available    | OS       |RNA       | [26997480](https://pubmed.ncbi.nlm.nih.gov/26997480/)| 
| Hwang       | Discovery  | 21            | Lung         | Available    | PFS/OS   |RNA       | [31959763](https://pubmed.ncbi.nlm.nih.gov/31959763/)| 
| Jerby_Arnon | Discovery  | 112           | Melanoma     | Available    | PFS      |RNA       | [30388455](https://pubmed.ncbi.nlm.nih.gov/30388455/)| 
| Jung        | Discovery  | 27            | Lung         | Available    | PFS      |RNA/DNA   | [31537801](https://pubmed.ncbi.nlm.nih.gov/31537801/)| 
| Kim         | Validation | 45            | Gastric      | Available    | None     |RNA       | [30013197](https://pubmed.ncbi.nlm.nih.gov/30013197/)| 
| Limagne1    | Discovery  | 70            | Lung         | Available    | PFS      |RNA       | [35051357](https://pubmed.ncbi.nlm.nih.gov/35051357/)| 
| Limagne2    | Discovery  | 26            | Lung         | Available    | PFS      |RNA       | [35051357](https://pubmed.ncbi.nlm.nih.gov/35051357/)| 
| Liu         | Discovery  | 121           | Melanoma     | Available    | PFS/OS   |RNA/DNA   | [31792460](https://pubmed.ncbi.nlm.nih.gov/31792460/)| 
| Miao1       | Discovery  | 33            | Kidney       | Available    | PFS/OS   |RNA/DNA   | [29301960](https://pubmed.ncbi.nlm.nih.gov/29301960/)|
| Nathanson   | Discovery  | 24            | Melanoma     | Available    | OS       |RNA/DNA   | [27956380](https://pubmed.ncbi.nlm.nih.gov/27956380/)|
| Padron      | Validation | 45            | Pancreas     | Available    | PFS/OS   |RNA       | [35662283](https://pubmed.ncbi.nlm.nih.gov/35662283/)|
| Puch        | Validation | 55            | Melanoma     | Available    | None     |RNA       | [32472114](https://pubmed.ncbi.nlm.nih.gov/32472114/)|
| Riaz        | Discovery  | 46            | Melanoma     | Available    | OS       |RNA/DNA   | [29033130](https://pubmed.ncbi.nlm.nih.gov/29033130/)|
| Snyder      | Discovery  | 25            | Ureteral     | Available    | PFS/OS   |RNA       | [28552987](https://pubmed.ncbi.nlm.nih.gov/28552987/)|
| Van_Allen   | Discovery  | 42            | Melanoma     | Available    | PFS/OS   |RNA/DNA   | [26359337](https://pubmed.ncbi.nlm.nih.gov/26359337/)|
| VanDenEnde  | Validation | 35            | Esophageal   | Available    | None     |RNA       | [33504550](https://pubmed.ncbi.nlm.nih.gov/33504550/)|
| Mariathasan | Discovery  | 195           | Bladder      | Available    | OS       |RNA/DNA   | [29443960](https://pubmed.ncbi.nlm.nih.gov/31792460/)| 
|             | Discovery  | 67            | Kidney       |                                                      | 
|             | Discovery  | 26            | Lymph_node   |                                                      | 
|             | Discovery  | 26            | Ureteral     |                                                      |
| Total       |            | 1355          | Pan-cancer   | [GitHub](https://github.com/bhklab/ClinSets.git)     |

: (\#tab:table1) Detailed overview for the immunotherapy datasets.

Table \@ref(tab:table1) shows an overview of the immunotherapy datasets and the total patients (n=1355)[CHECK]. The immunotherapy dataset is stored in a SummarizedExperiment object. For RNA profiles, we utilize log2-transformed TPM (Transcripts Per Million) data from protein-coding genes. For each dataset, genes with zero expression value in at least 50\% of the samples are filtered out. Note that only studies with at least 20 patients are included in the following analyses.

As an example, the Braun immunotherapy dataset with PMID [32472114](https://pubmed.ncbi.nlm.nih.gov/32472114/) includes RNA expression, clinical characteristics, and gene meta data. 181 patients have both clinical data and RNA expression data. In addition, the Braun dataset has 18,808 protein-coding genes. 

```{r load braun data}

# gene expression, clinical characteristics, and gene meta data 
load(file.path(dir, "ICB_Braun__Kidney__PD-(L)1.rda")) 
expr <- assay(dat_icb)
clin <- colData(dat_icb) %>% as.data.frame()
annot <- rowData(dat_icb) %>% as.data.frame()

# number of protein-coding genes and patients
dim(expr)

```

The detailed clinical characteristics of the Braun dataset including cancer type, age, sex, response (R vs. NR), overall survival (OS), and progression-free survival (PFS). The summary table represents that in the Braun dataset about 22\% of the response data is missing, while no missing for survival variables. Additionally, more than 80\% of patients are female. Note that "R" and "NR" represent responder and non-responder, respectively.  

```{r Braun clinical data summary}

# summary table of selected clinical variable 
sub_clin <- clin[, c("age","sex", "response","survival_time_os", "survival_time_pfs")]

print(dfSummary(sub_clin, 
                style = "grid", 
                plain.ascii = FALSE, 
                graph.col = FALSE), 
      method = 'render')

```


# Load Immunotherapy RNA Signatures 

We evaluate the reproducibility of a compendium of about 40 RNA signatures described as immunotherapy biomarkers. The following published immunotherapy RNA signatures are available on [GitHub](https://github.com/bhklab/SignatureSets) data repository. The signatures are also obtained from a web application [predictio.ca](https://predictio.ca/) to allow the community to easily mine our large compendium of published signatures.

```{r signature data}

app_dir <- str_split(rstudioapi::getActiveDocumentContext()$path,'PredictioR')[[1]][1]
dir_GeneSig <- file.path(app_dir, 'PredictioR', 'data/signature')
dir_GeneSig_info <- file.path(app_dir, 'PredictioR', 'data-raw')

signature <- read.csv( file.path(dir_GeneSig_info, "signature_INFO.csv")) 
signature$Signature <- as.character( signature$Signature )
signature$method <- as.character( signature$method )

```

|Signature  name|Method |PMID |      
  |-----------------|------------------------|------|
  | ADO             | GSVA                   | [31953314](https://pubmed.ncbi.nlm.nih.gov/31953314/) |
  | APM_Thompson    | GSVA                   | [33028693](https://pubmed.ncbi.nlm.nih.gov/33028693/) |
  | APM_Wang        | GSVA                   | [31767055](https://pubmed.ncbi.nlm.nih.gov/31767055/) |
  | B_cell_Budczies | GSVA                   | [33520406](https://pubmed.ncbi.nlm.nih.gov/33520406/) |
  | B_cell_Helmink  | GSVA                   | [31942075](https://pubmed.ncbi.nlm.nih.gov/31942075/) |
  | Blood_Response  | GSVA                   | [28807052](https://pubmed.ncbi.nlm.nih.gov/28807052/) |
  | CD8_Jiang       | GSVA                   | [30127393](https://pubmed.ncbi.nlm.nih.gov/30127393/) |
  | COX_IS          | specific algorithm     | [33220234](https://pubmed.ncbi.nlm.nih.gov/33220234/) |
  | IPRES           | specific algorithm     | [31683225](https://pubmed.ncbi.nlm.nih.gov/31683225/) |
  | C_ECM           | weighted mean          | [30410077](https://pubmed.ncbi.nlm.nih.gov/30410077/) |
  | EMT_Thompson    | weighted mean          | [31683225](https://pubmed.ncbi.nlm.nih.gov/31683225/) |
  | PTEN_MITF       | weighted mean          | [32245160](https://pubmed.ncbi.nlm.nih.gov/32245160/) |
  | T_cell_exclusion| weighted mean          | [30388455](https://pubmed.ncbi.nlm.nih.gov/30388455/) |
  | CD8_SF          | GSVA                   | [30388456](https://pubmed.ncbi.nlm.nih.gov/30388456/) |
  | CRMA            | GSVA                   | [29656892](https://pubmed.ncbi.nlm.nih.gov/29656892/) |
  | CYT             | GSVA                   | [25594174](https://pubmed.ncbi.nlm.nih.gov/25594174/) |
  | IFNG            | GSVA                   | [28650338](https://pubmed.ncbi.nlm.nih.gov/28650338/) |
  | ImmuneCells     | GSVA                   | [33033253](https://pubmed.ncbi.nlm.nih.gov/33033253/) |
  | Inflammatory    | GSVA                   | [31683225](https://pubmed.ncbi.nlm.nih.gov/31683225/) |
  | IRG_Ayers       | GSVA                   | [28650338](https://pubmed.ncbi.nlm.nih.gov/28650338/) |
  | KDM5A           | GSVA                   | [32908002](https://pubmed.ncbi.nlm.nih.gov/32908002/) |
  | LRRC15_CAF      | GSVA                   | [31699795](https://pubmed.ncbi.nlm.nih.gov/31699795/) |
  | M1              | GSVA                   | [31959763](https://pubmed.ncbi.nlm.nih.gov/31959763/) |
  | Macrophage      | GSVA                   | [33520406](https://pubmed.ncbi.nlm.nih.gov/33520406/) |
  | Myeloid_DC      | GSVA                   | [31942075](https://pubmed.ncbi.nlm.nih.gov/31942075/) |
  | NonResponse_Ivy | GSVA                   | [32907939](https://pubmed.ncbi.nlm.nih.gov/32907939/) |
  | PDL1            | GSVA                   | [30127393](https://pubmed.ncbi.nlm.nih.gov/30127393/) |
  | peri_Tcell      | GSVA                   | [31959763](https://pubmed.ncbi.nlm.nih.gov/31959763/) |
  | Response_Ivy    | GSVA                   | [32907939](https://pubmed.ncbi.nlm.nih.gov/32907939/) |
  | STAT1           | GSVA                   | [31316010](https://pubmed.ncbi.nlm.nih.gov/31316010/) |
  | T_cell_inflamed | GSVA                   | [30804515](https://pubmed.ncbi.nlm.nih.gov/30804515/) |
  | TGFB_Mariathasan| GSVA                   | [29443960](https://pubmed.ncbi.nlm.nih.gov/29443960/) |
  | TIS             | GSVA                   | [31684954](https://pubmed.ncbi.nlm.nih.gov/31684954/) |
  | TLS             | GSVA                   | [31942071](https://pubmed.ncbi.nlm.nih.gov/31942071/) |
  | IPS             | specific algorithm     | [28052254](https://pubmed.ncbi.nlm.nih.gov/28052254/) |
  | IMPRES          | weighted mean          | [30127394](https://pubmed.ncbi.nlm.nih.gov/30127394/) |
  | IRG_Yang        | weighted mean          | [31741756](https://pubmed.ncbi.nlm.nih.gov/31741756/) |
  | MPS             | weighted mean          | [32284588](https://pubmed.ncbi.nlm.nih.gov/32284588/) |
  | PredictIO       | specific algorithm     | [36055464](https://pubmed.ncbi.nlm.nih.gov/36055464/)|
  
  : (\#tab:table2) Detailed overview for the RNA immunotherapy signatures.
     
Table \@ref(tab:table2) shows an overview of signatures data and methods to compute the signatures' scores including weighted mean, Gene Set Variation Analysis (GSVA) [PMID 23323831](https://pubmed.ncbi.nlm.nih.gov/23323831/), and specific computational
algorithms.   

# Biomarkers and Immunotherapy Response Association

Association of specific biomarkers with immunotherapy response (R vs NR) is assessed using a logistic regression model. Association of specific biomarkers with immunotherapy survival (either PFS or OS) is assessed using Cox proportional hazards model to assess the discrimination or separation of a survival model. 

To assess the association between a biomarker and immunotherapy response (R vs NR), the logistic regression model is given as

$$\textit{logit}(E(Y|X)) = \textit{logit}(p) = \textit{ln} (\frac{p}{1-p}) = \beta_0 + \beta_1 X + \epsilon$$
where

* $Y$ denotes the outcome or dependent variable (e.g., immunotherapy response); this is a binary variable
* $X$ denotes the predictor of interest or the independent variable (e.g., gene expression)
* $p$ denotes the probability of the event occurring, and $\frac{p}{1-p}$ shows the odds ratio (OR)
* $\beta_0$ denotes the Y-intercept when X is zero; this is not informative for logistic regression models
* $\beta_1$ denotes the slope or the expected change in log odds per unit change in X
* $\epsilon$ denotes the random errors follow normal distribution

To assess the association between a biomarker and survival immunotherapy (e.g., OS or PFS), we fit the Cox proportional hazards (PH) model. The semi-parametric Cox PH model is given as 

$$h(t) = h_0(t) \times \text{exp}(\beta X)$$
where 

* $t$ represents the survival time.
* $h(t)$ is the hazard function determined by a set of covariate (e.g., X or gene expression)
* The coefficient $\beta$ measures the impact of covariate
* The term $h_0$ is called the baseline hazard. It corresponds to the value of the hazard if the covariate is equal to zero. The $t$ in $h(t)$ reminds us that the hazard may vary over time
* The quantity $\text{e}^\beta$ is called hazard ratios (HR)

When needed p-values are corrected for multiple testing using the Benjamini-Hochberg (False Discovery Rate, FDR) method. Associations are deemed statistically significant for p-values and/or FDR lower or equal to 5\%.

Note that Coef represents the log hazard ratio (logHR) or log odds ratio (logOR) depending on the analysis. 

## Association of Gene (CXCL9) with Immunotherapy Overall Survival Response 

We limit the association of CXCL9 for a selected Pancreas dataset (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/31792460/)). We assess the association of CXCL9 gene with immunotherapy survival response (OS), to gain more insight into the mechanism of response and resistance to immunotherapy response. The results show that CXCL9 is significantly associated with low survival (logHR = -3.47, p = 3e-05). Note the HR represents the logHR.   

```{r gene association os CXCL9, message=FALSE}

app_dir <- str_split(rstudioapi::getActiveDocumentContext()$path,'PredictioR')[[1]][1]
                     
source(file.path(app_dir, 'PredictioR', "R/getHR.R"))
source(file.path(app_dir, 'PredictioR', "R/getSummarizedExperiment.R"))
source(file.path(app_dir, 'PredictioR', "R/getGeneAssociation.R"))
source(file.path(app_dir, 'PredictioR', "R/getGeneSigAssociation.R"))
source(file.path(app_dir, 'PredictioR', "R/getMetaAnalysis.R"))
source(file.path(app_dir, 'PredictioR', "R/getGeneSigScore.R"))

load(file.path(dir_GeneSig, "PredictIO.rda"))
genes <- sig$gene_name

## merge the immunotherapy expression clinical data as a list
expr <- lapply(1:length(list.files(dir)), function(k){ 
  load(file.path(dir, list.files(dir)[k])) 
  dat_icb
})

study_icb <- substr(list.files(dir), 5, nchar(list.files(dir)) - 4)
names(expr) <- study_icb

group <- c("Discovery", "Validation", "Discovery", "Discovery", "Discovery", "Discovery",
           "Validation", "Discovery", "Discovery", "Discovery", "Discovery", "Discovery",
           "Discovery", "Discovery", "Discovery", "Discovery", "Discovery", "Validation", 
           "Validation", "Discovery", "Discovery", "Discovery", "Discovery", "Validation")

## Association of CXCL9 gene with OS for selected dataset Mariathasan Bladder

dat_expr <- as.matrix(assay(expr[["Mariathasan__Bladder__PD-(L)1"]]))
dat_clin <- colData(expr[["Mariathasan__Bladder__PD-(L)1"]]) %>% as.data.frame()
cancer_type <- names( table( dat_clin$cancer_type )[ table( dat_clin$cancer_type ) >= 20 ] )
       
survCont( surv = dat_clin$event_occurred_os[ dat_clin$cancer_type %in% cancer_type ] ,
          time = dat_clin$survival_time_os[ dat_clin$cancer_type %in% cancer_type ] ,
          time.censor = 36 , 
          var = as.numeric( scale( dat_expr["CXCL9" , ] )))
     
```

## Association of Gene (CXCL9) with Immunotherapy Progression-free Survival Response 

We limit the association of CXCL9 for a selected Pancreas dataset (Padron, [PMID 35662283](https://pubmed.ncbi.nlm.nih.gov/35662283/)). The default median cutoff is applied to stratify the continuous expression to the High and Low groups. We assess the association of CXCL9 gene with immunotherapy progression-free survival response (PFS), to gain more insight into the mechanism of response and resistance to immunotherapy response. The results show that CXCL9 is significantly associated with low survival (logHR = -0.72, p = 0.03). Note the HR represents the logHR.     


```{r gene association pfs CXCL9 dicho, message=FALSE}

dat_expr <- as.matrix(assay(expr[["Padron__Pancreas__PD-(L)1"]]))
dat_clin <- colData(expr[["Padron__Pancreas__PD-(L)1"]]) %>% as.data.frame()
cancer_type <- names( table( dat_clin$cancer_type )[ table( dat_clin$cancer_type ) >= 20 ] )
       
survDicho( surv = dat_clin$event_occurred_pfs[ dat_clin$cancer_type %in% cancer_type ] ,
           time = dat_clin$survival_time_pfs[ dat_clin$cancer_type %in% cancer_type ] ,
           time.censor= 24,
           var = as.numeric( scale( dat_expr["CXCL9" , ] )),
           n0.cutoff = 5,
           n1.cutoff = 5,
           method = "median")

```

### Kaplan-Meier Plot

The log-rank test is the most widely non-parametric used method of comparing two survival curves can be applied to assess difference in survival between the two Low and High groups.

```{r Fig1, message=FALSE, fig.width=5, fig.height=4, fig.cap = "CXCL9 association with immunotherapy PFS", fig.align="center"}
  
KMPlot(surv = dat_clin$event_occurred_pfs[ dat_clin$cancer_type %in% cancer_type ],
       time = dat_clin$survival_time_pfs[ dat_clin$cancer_type %in% cancer_type ],
       time.censor = 24,
       var =  as.numeric(scale(dat_expr["CXCL9", ])), 
       title = "CXCL9 and PFS Association",
       xlab = "Time (Months)",
       ylab = "Progression-free Survival", 
       n0.cutoff = 5, 
       n1.cutoff = 5)

```


## Association of Genes with Immunotherapy Overall Survival Response

We assess the association of 100 genes in PredictIO signature with immunotherapy survival (OS), to gain more insight into the mechanism of response and resistance to immunotherapy response. 

```{r gene association os , message=FALSE}

## identify the genes' association with OS
cox_os <- lapply(1:length(study_icb), function(i){
       
       geneSurvCont(dat.icb = expr[[i]],
                    time.censor = 36,
                    missing.perc = 0.5,
                    n.cutoff = 20,
                    feature = genes,
                    study = paste(strsplit(names(expr)[i], "__")[[1]][1], 
                                  strsplit(names(expr)[i], "__")[[1]][2],
                                  sep="__"),
                    surv.outcome = "OS")
  
  })
     
cox_os <- do.call(rbind, cox_os)
cox_os <- cox_os[!is.na(cox_os$Gene), ]
     
knitr::kable(head(cox_os[order(cox_os$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```


The association results for a selected Melanoma (Van Allen, [PMID 26359337](https://pubmed.ncbi.nlm.nih.gov/26359337/)) dataset show that 36\% of genes are associated with immunotherapy survival with $FDR < 0.05$, where about 92\% of them are associated with worse survival.     
     
```{r Fig2, message=FALSE, fig.width=5, fig.height=4, fig.cap = "gene association with immunotherapy overall survival", fig.align="center"}
     
df <- cox_os[cox_os$Study == "Van_Allen__Melanoma", ]
p <- volcanoPlot(feature = df$Gene, 
                 coef = df$Coef, 
                 pval = df$Pval, 
                 padj = df$FDR, 
                 neg.cutoff = 33, 
                 pos.cutoff = 3, 
                 x.lab = "logHR estimate",
                 padj.label = TRUE)
     
p
     
```
     
Additionally, we assess the association of genes in PredictIO signature with immunotherapy response (R vs NR).
     
```{r gene association response, message=FALSE}
     
## identify genes' association with binary response (R vs NR)
logreg <- lapply(1:length(study_icb), function(i){
       
       geneLogReg(dat.icb = expr[[i]],
                  missing.perc = 0.5,
                  n.cutoff = 20,
                  feature = genes,
                  study = paste(strsplit(names(expr)[i], "__")[[1]][1], 
                                  strsplit(names(expr)[i], "__")[[1]][2],
                                  sep="__"))
  
 })
     
logreg <- do.call(rbind, logreg)
logreg <- logreg[!is.na(logreg$Gene), ]
     
knitr::kable(head(logreg[order(logreg$FDR, decreasing = FALSE), ]), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```
     
The association results for a selected Bladder (Mariathasan,[PMID 29443960](https://pubmed.ncbi.nlm.nih.gov/31792460/)) show that 35\% of genes are associated with immunotherapy response with $p <0.05$, while 26\% of them are associated with better respond.    
     
```{r Fig3, message=FALSE, fig.width=5, fig.height=4, fig.cap = "gene association with immunotherapy response (responder vs non-responder)", fig.align="center"}
     
df <- logreg[logreg$Study == "Mariathasan__Bladder", ]
p <- volcanoPlot(feature = df$Gene, 
                 coef = df$Coef, 
                 pval = df$Pval, 
                 padj = df$FDR, 
                 neg.cutoff = 26, 
                 pos.cutoff = 9, 
                 x.lab = "logOR estimate",
                 padj.label = FALSE)
     
p
     
```

### Aggregating Associations through Meta-analysis (Pan-cancer)
     
To improve reproducibility, the results of individual independent datasets are pooled using random-effects meta-analysis with inverse variance weighting in \textit{DerSimonian and Laird} random-effects models. Heterogeneity across studies is evaluated by using the $Q$ statistic along with $I^2$ index, which describes the total variation across studies attributable to heterogeneity rather than sampling error. Note that $I^2$ value of $>50\%$ along with Cochran’s $Q$ statistic $p < 0.05$ represents moderate-to-high heterogeneity. 

As an example, for a gene CXCL9, to generalize the association with immunotherapy survival, we apply a meta-analysis approach to integrate findings across datasets for pan-cancer analysis. The result shows that the gene is considerably associated with worse survival ($\text{logHR} = -0.2, p = 0.001$), where the significant heterogenity across datasets ($I^2 = 44\%$, $Q$ $p = 0.04$).  
     
     
```{r meta-analysis pan-cancer}
     
# meta-analysis for a gene across datasets
res_meta <- meta.fun(coef = cox_os[cox_os$Gene == "CXCL9", "Coef"], 
                     se = cox_os[cox_os$Gene == "CXCL9", "SE"],
                     study  = cox_os[cox_os$Gene == "CXCL9", "Study"], 
                     pval = cox_os[cox_os$Gene == "CXCL9", "Pval"], 
                     n = cox_os[cox_os$Gene == "CXCL9", "N"],
                     cancer.type = do.call(rbind, strsplit(cox_os[cox_os$Gene == "CXCL9", "Study"], split='__', fixed=TRUE))[, 2],
                     treatment = cox_os[cox_os$Gene == "CXCL9", "Treatment"],
                     feature = "CXCL9", 
                     cancer.spec = FALSE, 
                     treatment.spec = FALSE)
     
# meta-analysis results
res_meta$meta_output
     
# summary of meta-analysis results
knitr::kable(res_meta$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```
     
The following forest plot displays the aggregated results across datasets for pan-cancer analysis. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed.  Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.
     
```{r Fig4, fig.width=8, fig.height=5, fig.cap ="Forest plot (CXCL9): Pan-cancer analysis", fig.align= 'center'}
     
forestPlot(coef = cox_os[cox_os$Gene == "CXCL9", "Coef"], 
           se = cox_os[cox_os$Gene == "CXCL9", "SE"], 
           study  = cox_os[cox_os$Gene == "CXCL9", "Study"], 
           pval = cox_os[cox_os$Gene == "CXCL9", "Pval"], 
           n = cox_os[cox_os$Gene == "CXCL9", "N"], 
           cancer.type = do.call(rbind, strsplit(cox_os[cox_os$Gene == "CXCL9", "Study"], split='__', fixed=TRUE))[, 2],
           treatment = cox_os[cox_os$Gene == "CXCL9", "Treatment"],
           feature = "CXCL9", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```
     
### Aggregating Associations through Meta-analysis (Per-cancer)

For cancer-specific analysis, consider meta-analysis when there are at least 3 datasets. The following forest plot displays the aggregated results across datasets as cancer-specific analysis. 
     
```{r meta-analysis per-cancer}
     
# cancer specific meta-analysis for a gene across datasets
res_meta <- metaPerCan.fun(coef = cox_os[cox_os$Gene == "CXCL9", "Coef"], 
                           se = cox_os[cox_os$Gene == "CXCL9", "SE"],
                           study  = cox_os[cox_os$Gene == "CXCL9", "Study"], 
                           pval = cox_os[cox_os$Gene == "CXCL9", "Pval"], 
                           n = cox_os[cox_os$Gene == "CXCL9", "N"],
                           cancer.type = do.call(rbind, strsplit(cox_os[cox_os$Gene == "CXCL9", "Study"], split='__', fixed=TRUE))[, 2],
                           treatment = cox_os[cox_os$Gene == "CXCL9", "Treatment"],
                           feature = "CXCL9", 
                           cancer.spec = TRUE)
     
# summary of meta-analysis results for each cancer types
knitr::kable(rbind(res_meta$Kidney$meta_summery,
                   res_meta$Melanoma$meta_summery,
                   res_meta$Other$meta_summery), format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across Melanoma datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.

```{r Fig5a, fig.width=8, fig.height=9, fig.cap ="Forest plot (CXCL9): cancer-specific analysis", fig.align= 'center'}

## Melanoma cancer specific
dat <- res_meta$Melanoma$input_data

forestPlot(coef = dat[, "Coef"], 
           se = dat[, "SE"], 
           study  = dat[, "Study"], 
           pval = dat[, "Pval"], 
           n = dat[, "N"], 
           cancer.type = do.call(rbind, strsplit(dat[, "Study"], split='__', fixed=TRUE))[, 2],
           treatment = dat[, "Treatment"],
           feature = "CXCL9", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```

Subgroup meta-analysis analysis is considered to assess the impact of cancer type on the source of moderate-to-high heterogeneity. 

```{r Fig5b, fig.width=8, fig.height=9, fig.cap ="Forest plot (CXCL9): cancer-specific analysis", fig.align= 'center'}

forestPlotPerCan(coef = cox_os[cox_os$Gene == "CXCL9", "Coef"], 
                 se = cox_os[cox_os$Gene == "CXCL9", "SE"], 
                 study  = cox_os[cox_os$Gene == "CXCL9", "Study"], 
                 pval = cox_os[cox_os$Gene == "CXCL9", "Pval"], 
                 n = cox_os[cox_os$Gene == "CXCL9", "N"], 
                 cancer.type = do.call(rbind, strsplit(cox_os[cox_os$Gene == "CXCL9", "Study"], split='__', fixed=TRUE))[, 2],
                 treatment = cox_os[cox_os$Gene == "CXCL9", "Treatment"],
                 feature = "CXCL9", 
                 xlab = "logHR estimate", 
                 label = "logHR")
     
```


## Association of Binary Gene with Immunotherapy Progression-free Survival Response

We assess the association of CXCL9 gene with immunotherapy survival (PFS), to gain more insight into the mechanism of response and resistance to immunotherapy response. Median cutoff is applied to stratify the continuous expression to the High and Low groups. 

```{r gene association pfs, message=FALSE}

cox_pfs <- lapply(1:length(study_icb), function(i){
       
       geneSurvDicho(dat.icb = expr[[i]],
                     time.censor = 24,
                     missing.perc = 0.5,
                     n.cutoff = 20,
                     feature = "CXCL9",
                     study =  paste(strsplit(names(expr)[i], "__")[[1]][1], 
                                  strsplit(names(expr)[i], "__")[[1]][2],
                                  sep="__"),
                     surv.outcome = "PFS",
                     n1.cutoff = 5,
                     n0.cutoff = 5,
                     method = "median")
     })
     
cox_pfs <- do.call(rbind, cox_pfs)
cox_pfs <- cox_pfs[!is.na(cox_pfs$Gene), ]
     
knitr::kable(cox_pfs, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()


```

### Aggregating Associations through Meta-analysis (Pan-cancer)

As an example, for a gene CXCL9, to generalize the association with immunotherapy PFS, we apply a meta-analysis approach to integrate findings across datasets for pan-cancer analysis. The result shows that the gene is considerably associated with worse survival ($\text{logHR} = -0.47, p = 0.002$), where the significant heterogenity across datasets ($I^2 = 55\%$, $Q$ $p = 0.01$).  

```{r meta-analysis pan-cancer dicho}
     
# meta-analysis for a gene across datasets
res_meta <- meta.fun(coef = cox_pfs[, "Coef"], 
                     se = cox_pfs[, "SE"],
                     study  = cox_pfs[, "Study"], 
                     pval = cox_pfs[, "Pval"], 
                     n = cox_pfs[, "N"],
                     cancer.type = do.call(rbind, strsplit(cox_pfs[, "Study"], split='__', fixed=TRUE))[, 2],
                     treatment = cox_pfs[, "Treatment"],
                     cancer.spec = FALSE,
                     treatment.spec = FALSE,
                     feature = "CXCL9")
     
# summary of meta-analysis results
knitr::kable(res_meta$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across datasets for pan-cancer analysis. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed.  Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.

```{r Fig6, fig.width=8, fig.height=5, fig.cap ="Forest plot (CXCL9, binary median cutoff): Pan-cancer analysis", fig.align= 'center'}
     
forestPlot(coef = cox_pfs[ , "Coef"], 
           se = cox_pfs[ , "SE"], 
           study  = cox_pfs[ , "Study"], 
           pval = cox_pfs[ , "Pval"], 
           n = cox_pfs[ , "N"], 
           cancer.type = do.call(rbind, strsplit(cox_pfs[, "Study"], split='__', fixed=TRUE))[, 2],
           treatment = cox_pfs[, "Treatment"],
           feature = "CXCL9", 
           xlab = "logHR estimate", 
           label = "logHR")
     
     
```

### Aggregating Associations through Meta-analysis (Per-treatment)

For treatment-specific analysis, consider meta-analysis when there are at least 3 datasets. The following forest plot displays the aggregated results across datasets as treatment-specific analysis. 

```{r meta-analysis per-treatment dicho}
     
# meta-analysis for a gene across datasets
res_meta <- metaPerTreatment.fun(coef = cox_pfs[, "Coef"], 
                                 se = cox_pfs[, "SE"],
                                 study  = cox_pfs[, "Study"], 
                                 pval = cox_pfs[, "Pval"], 
                                 n = cox_pfs[, "N"],
                                 cancer.type = do.call(rbind, strsplit(cox_pfs[, "Study"], split='__', fixed=TRUE))[, 2],
                                 treatment = cox_pfs[, "Treatment"],
                                 treatment.spec = TRUE,
                                 feature = "CXCL9")
     
# summary of meta-analysis results
knitr::kable(res_meta$`PD-1/PD-L1`$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
    
```

The following forest plot displays the aggregated results across PD-1/PD-L1 datasets. For each dataset, the logHR, 95\% confidence intervals (CIs), and p-values are computed. Horizontal bars represent the 95\% CIs of logHR. The blue diamond represents the aggregated results.

```{r Fig7a, fig.width=8, fig.height=9, fig.cap ="Forest plot (CXCL9, binary median cutoff): treatment-specific analysis", fig.align= 'center'}

dat <- res_meta$`PD-1/PD-L1`$input_data   
forestPlot(coef = dat[ , "Coef"], 
           se = dat[ , "SE"], 
           study  = dat[ , "Study"], 
           pval = dat[ , "Pval"], 
           n = dat[ , "N"],
           cancer.type = do.call(rbind, strsplit(dat[, "Study"], split='__', fixed=TRUE))[, 2],
           treatment = dat[, "Treatment"],
           feature = "CXCL9", 
           xlab = "logHR estimate", 
           label = "logHR")
     
```


## Signature Score Computation
     
RNA expression signatures including genes without any weight are computed using Gene Set Variation Analysis (GSVA) enrichment score using the \textit{GSVA} R package [REF]. RNA expression signatures containing genes associated with specific weights (+1 and -1 for up and down-regulated genes, respectively) are computed using the weighted mean expression approach. For RNA immunotherapy signatures with specific computational algorithms to get scores such as PredictIO, we can use their original publication to get the signature score. 
     
For each dataset, RNA signatures are only computed if at least $80\%$ of their genes are present in the dataset. We also apply \textit{z}-score transformation to the genes of each RNA signature before the computation. The \textit{z}-score transformation is applied before GSVA or the weighted mean computation and after the computation of the specific immunotherapy signatures.
     
### GSVA signature score computation
     
\textit{APM_Thompson} ([PMID 33028693](https://pubmed.ncbi.nlm.nih.gov/33028693/)) is a RNA signature capturing the antigen processing and presentation machinery (APM) activity. This APM signature was a predictor of the response of PD-(L)1 blockade in Lung and Melanoma patients. To get the signature score, the GSVA approach is applied. 
     
```{r signature score computation GSVA, message=FALSE, warning = FALSE, fig.width=9, fig.height=5.5}
     
i<- which(signature$Signature == "APM_Thompson")
load(file.path(dir_GeneSig, "APM_Thompson.rda"))
     
geneSigScore <- lapply(1:length(expr), function(k){
       
       geneSig <- geneSigGSVA(dat.icb = expr[[k]], 
                              sig = sig,
                              sig.name = signature$Signature[i],
                              missing.perc = 0.5,
                              n.cutoff = 20,
                              sig.perc = 0.8, 
                              study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                            strsplit(names(expr)[k], "__")[[1]][2],
                                            sep="__"))
       
       data.frame(study = names(expr)[k],
                  signature_name = signature$Signature[i], 
                  geneSig)
       
     })
     
geneSigScore <- do.call(rbind, geneSigScore)
geneSigScore <- geneSigScore[!is.na(geneSigScore$geneSig), ]  
     
```
     
The box plots denote the distribution of computed signature score across immunotherapy datasets. 
     
```{r Fig8, message=FALSE, warning = FALSE, fig.width=10, fig.height=6, fig.cap= "Distribution of APM_Thompson signature score across datasets using GSVA approach.", fig.align= "center"}
     
ggplot(geneSigScore, aes(x = study, y = geneSig)) + 
       geom_boxplot(fill= "#fdb863", color = "#878787") +
       xlab("") +
       ylab("signature score (GSVA)") +
       theme(axis.text.x=element_text(size=10,  face="bold", angle = 90),
             axis.title=element_text(size=12,face="bold"),
             axis.text.y=element_text(size=10, face = "bold"),
             panel.grid.major = element_blank(), 
             panel.grid.minor = element_blank(),
             panel.background = element_blank(),
             plot.background = element_blank(), 
             axis.line = element_line(colour = "black"),
             legend.position="none",
             legend.text = element_text(size = 6, face="bold"),
             legend.title = element_blank()) 
     
     
  
```
     
     
### Weighted mean signature score computation
     
\textit{T_cell_exclusion} ([PMID 30388455](https://pubmed.ncbi.nlm.nih.gov/30388455/)) is a RNA signature capturing the mechanism of T-cell exclusion. This T_cell_exclusion signature was a predictor of the failure of PD-1 blockade in Melanoma patients. To get the signature score, the weighted mean approach is applied. 
     
```{r signature score computation weighted mean, message=FALSE, warning = FALSE}
     
i<- which(signature$Signature == "T_cell_exclusion")
load(file.path(dir_GeneSig, "T_cell_exclusion.rda"))
     
geneSigScore <- lapply(1:length(expr), function(k){
       
       geneSig <- geneSigMean(dat.icb = expr[[k]], 
                              sig = sig,
                              sig.name = signature$Signature[i],
                              missing.perc = 0.5,
                              n.cutoff = 20,
                              sig.perc = 0.8, 
                              study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                            strsplit(names(expr)[k], "__")[[1]][2],
                                            sep="__"))
       
       data.frame(study = names(expr)[k],
                  signature_name = signature$Signature[i], 
                  geneSig)
       
     })
     
geneSigScore <- do.call(rbind, geneSigScore)
geneSigScore <- geneSigScore[!is.na(geneSigScore$geneSig), ]  
    
```
     
The box plots denote the distribution of computed signature score across immunotherapy datasets. 
    
```{r Fig9, message=FALSE, warning = FALSE, fig.width=10, fig.height=6, fig.cap ="Distribution of IRG_Yang signature score across datasets using weighted mean approach", fig.align= 'center'}
     
ggplot(geneSigScore, aes(x = study, y = geneSig)) + 
       geom_boxplot(fill= "#a6dba0", color = "#878787") +
       xlab("") +
       ylab("signature score (weighted mean)") +
       theme(axis.text.x=element_text(size=10,  face="bold", angle = 90),
             axis.title=element_text(size=12,face="bold"),
             axis.text.y=element_text(size=10, face = "bold"),
             panel.grid.major = element_blank(), 
             panel.grid.minor = element_blank(),
             panel.background = element_blank(),
             plot.background = element_blank(), 
             axis.line = element_line(colour = "black"),
             legend.position="none",
             legend.text = element_text(size = 6, face="bold"),
             legend.title = element_blank()) 
     
     
```
     
### Specific algorithm (PredictIO) signature score computation
     
\textit{PredictIO} ([PMID 36055464](https://pubmed.ncbi.nlm.nih.gov/36055464/)) is a RNA signature that demonstrated better and more consistent ability to predict immunotherapy response as compared to other signatures. The specific algorithm is applied to computed the signature score. The PredictIO signature was identified by integrating gene expression data from a large pan-cancer discovery immunotherapy datasets. The validation immunotherapy datasets were used to assess the performance of the PredictIO compared with other published RNA immunotherapy signatures.
     
     
```{r signature score computation specific, message=FALSE, warning = FALSE}
     
i<- which(signature$Signature == "PredictIO")
load(file.path(dir_GeneSig, "PredictIO.rda"))
     
geneSigScore <- lapply(1:length(expr), function(k){
       
       geneSig <- geneSigPredictIO(dat.icb = expr[[k]], 
                                   sig = sig,
                                   sig.name = signature$Signature[i],
                                   missing.perc = 0.5,
                                   n.cutoff = 20,
                                   sig.perc = 0.8, 
                                   study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                                 strsplit(names(expr)[k], "__")[[1]][2],
                                                 sep="__"))
       
       data.frame(study = names(expr)[k],
                  status = group[k],
                  signature_name = signature$Signature[i], 
                  geneSig)
       
     })
     
geneSigScore <- do.call(rbind, geneSigScore)
geneSigScore <- geneSigScore[!is.na(geneSigScore$geneSig), ]  
     
```
     
The box plots denote the distribution of computed signature score across immunotherapy datasets. 
     
```{r Fig10, message=FALSE, warning = FALSE, fig.width=10, fig.height=6, fig.cap ="Distribution of PrdictIO signature score across datasets (discovery and validation) using specific algorithm", fig.align= 'center'}
     
ggplot(geneSigScore, aes(x = study, y = geneSig, fill = status)) + 
       geom_boxplot(color = "#878787") + 
       scale_fill_manual(values = c("#fee08b", "#abd9e9")) +
       xlab("") +
       ylab("signature score (specific algorithm)") +
       theme(axis.text.x=element_text(size=10,  face="bold", angle = 90),
             axis.title=element_text(size=12,face="bold"),
             axis.text.y=element_text(size=10, face = "bold"),
             panel.grid.major = element_blank(), 
             panel.grid.minor = element_blank(),
             panel.background = element_blank(),
             plot.background = element_blank(), 
             axis.line = element_line(colour = "black"),
             legend.position="right",
             legend.text = element_text(size = 6, face="bold"),
             legend.title = element_blank()) 
     
```
     
## Association of RNA Signatures with Immunotherapy Response
     
For \textit{APM_Thompson} signature, we assess the association of signature with immunotherapy response (PFS). 
    
```{r signature score association with immunotherapy pfs, message=FALSE, warning = FALSE}
     
i<- which(signature$Signature == "APM_Thompson")
load(file.path(dir_GeneSig, "APM_Thompson.rda"))
     
geneSig_pfs <- lapply(1:length(expr), function(k){
       
       geneSig <- geneSigGSVA(dat.icb = expr[[k]], 
                              sig = sig,
                              sig.name = signature$Signature[i],
                              missing.perc = 0.5,
                              n.cutoff = 20,
                              sig.perc = 0.8, 
                              study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                            strsplit(names(expr)[k], "__")[[1]][2],
                                            sep="__"))
       
       if(sum(!is.na(geneSig)) > 0){
         
         res <- geneSigSurvCont(dat.icb = expr[[k]],
                                geneSig = geneSig,
                                time.censor = 24,
                                n.cutoff = 20,
                                study =  paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                               strsplit(names(expr)[k], "__")[[1]][2],
                                               sep="__"),
                                surv.outcome = "PFS",
                                sig.name = signature$Signature[i])
         
       }else{
         
         res <- data.frame( Outcome = "PFS",
                            Gene = geneSig, 
                            Study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                          strsplit(names(expr)[k], "__")[[1]][2],
                                          sep="__"),
                            Coef = NA,
                            SE = NA,
                            N = NA,
                            Pval = NA,
                            Treatment = NA) 
       }
       
       rownames(res) <- NULL
       res
       
     })
     
geneSig_pfs <- do.call(rbind, geneSig_pfs)
geneSig_pfs <- geneSig_pfs[!is.na(geneSig_pfs$Coef), ]
     
# top association results
knitr::kable(geneSig_pfs[order(geneSig_pfs$Pval, decreasing = FALSE), ], format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
     
```
     
The associations across datasets are aggregated using the meta-analysis approach. The signature is negatively associated with PFS ($logHR = -0.60, p = 0.001$). 
    
```{r signature score meta-analysis pan, message=FALSE, warning = FALSE }
     
res <- meta.fun(coef = geneSig_pfs$Coef, 
                se = geneSig_pfs$SE, 
                study  = geneSig_pfs$Study, 
                pval = geneSig_pfs$Pval, 
                n = geneSig_pfs$N, 
                cancer.type = do.call(rbind, strsplit(geneSig_pfs$Study, split='__', fixed=TRUE))[, 2],
                treatment = geneSig_pfs$Treatment,
                feature = unique(geneSig_pfs$Gene),
                cancer.spec = FALSE,
                treatment.spec = FALSE)
     
     
# summary of meta-analysis
knitr::kable(res$meta_summery, format="html", 
                  table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()
     
```
     
The forest plot shows the pan-cancer analysis. 
     
```{r Fig11, message=FALSE, warning = FALSE, fig.width=8, fig.height=5, fig.cap ="Forest plot (APM_Thompson signature): cancer-specific analysis", fig.align= 'center'}
     
forestPlot(coef = geneSig_pfs$Coef, 
           se = geneSig_pfs$SE, 
           study  = geneSig_pfs$Study, 
           pval = geneSig_pfs$Pval, 
           n = geneSig_pfs$N, 
           cancer.type = do.call(rbind, strsplit(geneSig_pfs$Study, split='__', fixed=TRUE))[, 2],
           treatment = geneSig_pfs$Treatment,
           feature = unique(geneSig_pfs$Gene), 
           xlab = "logHR estimate", 
           label = "logHR")
     
     
```
     
The subgroup meta-analysis is also considered to assess the impact of cancer type on the source of moderate-to-high heterogeneity ($I^2=0.45$, $Q$ $p=0.06$). 
     
     
```{r Fig11a, message=FALSE, warning = FALSE, fig.width=8, fig.height=9, fig.cap ="Forest plot (APM_Thompson signature): cancer-specific analysis", fig.align= 'center'}
     
forestPlotPerCan(coef = geneSig_pfs$Coef, 
                 se = geneSig_pfs$SE, 
                 study  = geneSig_pfs$Study, 
                 pval = geneSig_pfs$Pval, 
                 n = geneSig_pfs$N, 
                 cancer.type = do.call(rbind, strsplit(geneSig_pfs$Study, split='__', fixed=TRUE))[, 2],
                 treatment = geneSig_pfs$Treatment,
                 feature = unique(geneSig_pfs$Gene), 
                 xlab = "logHR estimate", 
                 label = "logHR")
     
```

     
## Comparisons
     
Assess the association of RNA signatures with immunotherapy response (OS). Per signatures, we integrate the results across datasets to generalize the findings and improve the power of analyses. 

```{r immunotherapy signatures comparison, message=FALSE, warning = FALSE}

# remove PredictIO signatures from this analyses due to the nature of creating this signature using discovery datasets

GeneSig_list <- list.files(dir_GeneSig)[!(list.files(dir_GeneSig) %in% c("PredictIO.rda"))]
GeneSig_list <- GeneSig_list[order(GeneSig_list)]

signature <- signature[signature$Signature != "PredictIO", ]
signature <- signature[order(signature$Signature), ]

AllGeneSig_os <- lapply(1:length(GeneSig_list), function(i){  
  
  load(file.path(dir_GeneSig, GeneSig_list[i]))
  
  geneSig_os <- lapply(1:length(expr), function(k){
    
    if(signature$method[i] == "GSVA"){
      
      geneSig <- geneSigGSVA(dat.icb = expr[[k]], 
                             sig = sig,
                             sig.name = signature$Signature[i],
                             missing.perc = 0.5,
                             n.cutoff = 20,
                             sig.perc = 0.8, 
                             study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__")) 
      
    }     
    
    if(signature$method[i] == "weighted_mean"){
      
      geneSig <- geneSigMean(dat.icb = expr[[k]], 
                             sig = sig,
                             sig.name = signature$Signature[i],
                             missing.perc = 0.5,
                             n.cutoff = 20,
                             sig.perc = 0.8, 
                             study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__")) 
      
    }     
    
    if(signature$method[i] == "specific" & signature$Signature[i] == "COX_IS"){
      
      geneSig <- geneSigCOX_IS(dat.icb = expr[[k]], 
                               sig = sig,
                               sig.name = signature$Signature[i],
                               missing.perc = 0.5,
                               n.cutoff = 20,
                               sig.perc = 0.8, 
                               study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__"))  
      
    } 
    
    if(signature$method[i] == "specific" & signature$Signature[i] == "IPS"){
      
      geneSig <- geneSigIPS(dat.icb = expr[[k]], 
                            sig = sig,
                            sig.name = signature$Signature[i],
                            missing.perc = 0.5,
                            n.cutoff = 20,
                            sig.perc = 0.8, 
                            study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__")) 
      
    } 
    
    if(signature$method[i] == "specific" & signature$Signature[i] == "IPRES"){
      
      geneSig <- geneSigIPRES(dat.icb = expr[[k]], 
                              sig = sig,
                              sig.name = signature$Signature[i],
                              missing.perc = 0.5,
                              n.cutoff = 20,
                              sig.perc = 0.8, 
                              study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__"))  
      
    } 
    
    
    if(sum(!is.na(geneSig)) > 0){
      
      res <- geneSigSurvCont(dat.icb = expr[[k]],
                             geneSig = geneSig,
                             time.censor = 36,
                             n.cutoff = 20,
                             study =  paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__"),
                             surv.outcome = "OS",
                             sig.name = signature$Signature[i])
      
    }else{
      
      res <- data.frame( Outcome = "OS",
                         Gene = geneSig, 
                         Study = paste(strsplit(names(expr)[k], "__")[[1]][1], 
                                           strsplit(names(expr)[k], "__")[[1]][2],
                                           sep="__"),
                         Coef = NA,
                         SE = NA,
                         N = NA,
                         Pval = NA,
                         Treatment = NA) 
      
      
    }
    
    rownames(res) <- NULL
    res
    
  })
  
  geneSig_os <- do.call(rbind, geneSig_os)
  geneSig_os <- geneSig_os[!is.na(geneSig_os$Coef), ]
  rownames(geneSig_os) <- NULL
  geneSig_os  
  
})

AllGeneSig_meta <- lapply(1:length(AllGeneSig_os), function(j){
  
  res <- meta.fun(coef = AllGeneSig_os[[j]]$Coef, 
                  se = AllGeneSig_os[[j]]$SE, 
                  study  = AllGeneSig_os[[j]]$Study, 
                  pval = AllGeneSig_os[[j]]$Pval, 
                  n = AllGeneSig_os[[j]]$N, 
                  cancer.type = do.call(rbind, strsplit(AllGeneSig_os[[j]]$Study, split='__', fixed=TRUE))[, 2],
                  treatment = AllGeneSig_os[[j]]$Treatment,
                  cancer.spec = FALSE,
                  treatment.spec = FALSE,
                  feature = unique(AllGeneSig_os[[j]]$Gene))
  
  res$meta_summery 
})

AllGeneSig_meta <- do.call(rbind, AllGeneSig_meta)
AllGeneSig_meta$FDR <- p.adjust(AllGeneSig_meta$Pval, method = "BH")

# top signatures association results
knitr::kable(head(AllGeneSig_meta[order(AllGeneSig_meta$FDR, decreasing = FALSE), ]), format="html", 
             table.attr = "style='width:30%;'", row.names = FALSE) %>% kableExtra::kable_styling()

```

The volcano plot denotes the majority of signatures are associated with worse survival with FDR$< 0.05$.  


```{r Fig13, message=FALSE, warning = FALSE, fig.width=9, fig.height=5}

p1 <- volcanoPlot(feature = AllGeneSig_meta$Gene, 
                  coef = AllGeneSig_meta$Coef, 
                  pval = AllGeneSig_meta$Pval, 
                  padj = AllGeneSig_meta$FDR, 
                  neg.cutoff = 19, 
                  pos.cutoff = 1, 
                  x.lab = "logHR estimate",
                  padj.label = FALSE)


p2 <- volcanoPlot(feature = AllGeneSig_meta$Gene, 
                  coef = AllGeneSig_meta$Coef, 
                  pval = AllGeneSig_meta$Pval, 
                  padj = AllGeneSig_meta$FDR, 
                  neg.cutoff = 16, 
                  pos.cutoff = 1, 
                  x.lab = "logHR estimate",
                  padj.label = TRUE)

grid.arrange(p1, p2, ncol = 2)

```


# References


# SessionInfo

```{r sessionInfo, echo=FALSE}
#sessionInfo()
```

